### PR TITLE
[ClangModule] Improve the workaround

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Try to improve the workaround for ext4 file system.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please test llvm

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
Should this go to `next` branch first?  Does this change fix the Linux tests?

**Comment by cachemeifyoucan** [comment]
> Should this go to `next` branch first? Does this change fix the Linux tests?

Just FYI, not ready for landing yet. I think I might be able to fix more linux tests if not all. However, `next` fix might look a bit different because of all the optional FileEntry changes.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test llvm

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
Linux test failure:
```
  Clang :: ClangScanDeps/modules-include-tree-prefix-map.c
  LLVM :: Instrumentation/AddressSanitizer/global_with_comdat.ll
```
The second one is unrelated. The first one needs investigation but the issue doesn't seem to be module file loading related, but: 
```
Error while scanning dependencies for /home/build-user/llvm-project/build/tools/clang/test/ClangScanDeps/Output/modules-include-tree-prefix-map.c.tmp/dir1/tu.m:
/home/build-user/llvm-project/build/tools/clang/test/ClangScanDeps/Output/modules-include-tree-prefix-map.c.tmp/dir1/System/module.modulemap:3:10: error: header 'stdbool.h' not found
```

**Comment by cachemeifyoucan** [comment]
Verified `ClangScanDeps/modules-include-tree-prefix-map.c` is failing with the entire workaround reverted. The module map behavior is different on Ubuntu that `"stdbool.h"` is not found from module map file even it is a clang builtin header.

This is ready to be review/land.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test llvm

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test windows platform

File: clang/lib/Basic/FileManager.cpp [hunk 1]
```diff
@@ -472,9 +472,13 @@ FileEntryRef FileManager::getVirtualFileRef(StringRef Filename, off_t Size,
 }
 
 OptionalFileEntryRef FileManager::getBypassFile(FileEntryRef VF) {
+  // Compute the requested name and make it an absolute path.
+  SmallString<256> Path(VF.getNameAsRequested());
+  makeAbsolutePath(Path);
```
Comment (benlangmuir) [inline clang/lib/Basic/FileManager.cpp]: Why do we need `makeAbsolutePath` here?

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: There is a test to check a.pcm and /path/a.pcm both works, maybe `clang/test/Modules/explicit-build-relpath.cpp`
Ideally, this should be realpath, but that is going to be expensive. This is just to make the bad workaround less bad.

Comment (benlangmuir) [reply clang/lib/Basic/FileManager.cpp]: @jansvoboda11 any ideas here?

Comment (jansvoboda11) [reply clang/lib/Basic/FileManager.cpp]: That's curious. @cachemeifyoucan can you explain a bit more why this is necessary?

`FileManager::getFileRef()` does not make the requested path absolute and stores it as-is in the returned `FileEntryRef`. This means that we've been able to successfully use the relative path to obtain `VF`. Why exactly can't we use the same relative path (`VF.getNameAsRequested()`) here in `getBypassFile()`?

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: Didn't try and didn't investigate deep into the failure, just workaround the behavior difference.

My guess is B.pcm is built with relative path A.pcm, and A.pcm is also imported via absolute path. They need to be the same FileEntryRef to avoid duplicated module error.

Comment (jansvoboda11) [reply clang/lib/Basic/FileManager.cpp]: It will be hard to come up with better solution without knowing what exactly goes wrong in the test case. But in general, the intent of `FileManager::getBypassFile()` is to provide new `FileEntry` that's **not** de-duplicated based on the inode. If that's causing duplicate module errors down the line, it seems we'll need to fix the client.

If we make the path absolute here, we'll be returning a `FileEntryRef` that could potentially report different name compared to the original, which is surprising.

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: Current `getBypassFile` implementation is totally diverted from what the comments said. I really want to revert the workaround since the workaround is absolutely wrong, this change is just some middle ground that makes module work in more cases.

Comment (jansvoboda11) [reply clang/lib/Basic/FileManager.cpp]: How does it diverge? Do you mean the fact that it does de-duplicate bypass files based on `VF.getName()`?

What's your plan for reverting the workaround for ext4 recycling inodes?

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: getBypassFile is not supposed to do any dedup according to comments but it is already deduping. After the change, getBypassFile will only dedup from absolute path (not realpath), which is not unreasonable, or more unreasonable than the semantic before the patch.

I don't have plan to revert the change (just want to). All I can do is putting the workaround into a less horrible place.

Comment (benlangmuir) [reply clang/lib/Basic/FileManager.cpp]: Can we have the returned FileEntryRef use the requested path and only use the absolute path for dedup? That would be closer to how normal FileEntryRef lookup works.  I'm not so concerned about the dedup, I'm only concerned about us changing the path in the FileEntryRef.

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: The requested path is just key for dedup. I am not sure if I want to add another map for dedup only. I think we either keep makeAbsolute, or XFAIL that test.

Comment (cachemeifyoucan) [reply clang/lib/Basic/FileManager.cpp]: I decide to put the makeAbsolutePath behind the same guard so it is not changing behavior when module bypass is not used.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test llvm

**Comment by cachemeifyoucan** [comment]
@swift-ci please test windows platform

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test llvm

File: clang/lib/Basic/FileManager.cpp [hunk 2]
```diff
@@ -483,7 +483,7 @@ OptionalFileEntryRef FileManager::getBypassFile(FileEntryRef VF) {
 
   // If we've already bypassed just use the existing one.
   auto Insertion = SeenBypassFileEntries->insert(
-      {VF.getName(), std::errc::no_such_file_or_directory});
+      {VF.getNameAsRequested(), std::errc::no_such_file_or_directory});
```
Comment (benlangmuir) [inline clang/lib/Basic/FileManager.cpp:486]: Possibly this change should go upstream, but don't need to wait on that and not sure how to test it upstream.

**Review by benlangmuir — APPROVED** [review]

**Comment by cachemeifyoucan** [comment]
@swift-ci please test windows platform

**Comment by cachemeifyoucan** [comment]
@swift-ci please test macos platform

**Comment by drodriguez** [comment]
The comments of this PR talk about ext4 filesystems, but the compilation conditional check for `!APPLE`, which will get Windows into the mix and other filesystems that are not ext4. If the test suite is working correctly, Windows seems to be fine with these changes.

I am, however, seeing some problems in Linux during the test suite that go away if I comment out the parts introduced in this PR. I am testing on Linux, but my filesystem is btrfs, not ext4. I wonder if the check should be done only for certain filesystems, and not just for non-Apple platforms. This cannot be done in the preprocessor, and will need to be done during runtime, of course.

**Comment by cachemeifyoucan** [comment]
@drodriguez This is just an improvement for the previously broken workaround to make all the tests pass. All platforms and file systems will hit the same error if you revert this workaround. This is not an attempt to fix the underlying problem that the clang file manager is relying on the FileSystem inode behavior. That needs to be fixed in general in upstream LLVM and no one has the best idea for handling that yet.

We can't really distinguish file system on the file system because:
* There is no precedent for such behavior so there is lack of infrastructure to do that.
* There is potential to run into compilations that files are coming from different file system and it can't be handled by current design
* We can't check file system for each file access since that will be really expensive.

**Comment by drodriguez** [comment]
The "previously broken workaround" also misbehaved in our btrfs systems and we had it disabled for testing as well. Since this was being reworked I decided to ask to see if there was something that could be improved. Seems that Darwin systems do not need those extra bits of code, and seems that our btrfs systems do neither.

I looked around in the LLVM file system APIs and indeed they don't seem to have a way of distinguishing filesystems (which makes sense). I don't know if a more generic version of the fix work, one that works independently of the filesystem and do not break any of them.

One possibility might be adding code to either check an environment variable or a command line flag to enable/disable those bits. Use the defaults that work on your CI, but leave it open to other people to change the behaviour in their systems without modifying the code.

**Comment by cachemeifyoucan** [comment]
>  Seems that Darwin systems do not need those extra bits of code, and seems that our btrfs systems do neither.

That is not true. Darwin works only because it is `#if` out on Apple platform. If remove that `#if`, everything will fail before this commit.

The underlying problem on ext4 is that it recycles inode really fast on file deletion. It is generally safe to take this change if you live on apple branch regardless of your OS/FS or want to have a working implicit module build on a file system have similar file system inode implementation. The only behavior change comparing to upstream LLVM is that test I XFAIL'ed here which is really a corner case.

**Comment by drodriguez** [comment]
But that's the corner case we are seeing, and it does spill into some other assertions if we compile those `#if` in our version.

One of the system Clang modules (shipped with a Linux distribution) ends up being refered as two "different" files in the ModuleCache, one like `/path/to/ModuleCache/module.pcm` and another `/path/to/ModuleCache//module.pcm`. It still happens from an empty `ModuleCache` directory, so it is not something that can be avoided. I am not sure were the second `/` is being added, but it breaks the checks for identity (sorry I don't have the exact paths at hand, but the double directory separator was the clue).

Also, if those `#if` blocks are compiled, the assertions build fail in some tests (tests in `ClangScanDeps`, `Index`, `Modules` and `PCH` directories) with assertions like `missing module file for loaded module?` and `Invalid index into type index remap`, which happen just after calls into the `ModuleManager` and the `FileManager` that were modified by this code.

**Comment by cachemeifyoucan** [comment]
@benlangmuir @jansvoboda11 for comments. I do have a previous version that bypass the file as canonicalized path but that was shutdown for changing the behavior (clang file manager changes are kind of risky).

> Also, if those #if blocks are compiled, the assertions build fail in some tests (tests in ClangScanDeps, Index, Modules and PCH directories) with assertions like missing module file for loaded module? and Invalid index into type index remap, which happen just after calls into the ModuleManager and the FileManager that were modified by this code.

Are you talking about before this patch or after? Before is expected and what this patch aimed to fix. If it is after, please provide your configuration and error message so we can double check.

**Comment by drodriguez** [comment]
Before and after,

Before this patch, when the `#if !defined(__APPLE__)` was only in `ModuleManager::lookupModuleFile`, we needed to disable that bit to be able to build the standard library (because of that system Clang module). I did not found a test that did not pass under with the block commented out, but I cannot say if the tests would had fail because I was not able to compile the toolchain in the first place.

After this patch, I started only commenting out the same block in `lookupModuleFile`, which allowed me to compile, but then when checking Clang, I got around 30 failing tests, with the two different assertions I wrote before:

- `llvm-project/clang/lib/Serialization/ASTReader.cpp:7617: serialization::DeclID clang::ASTReader::getGlobalDeclID(ModuleFile &, LocalDeclID) const: Assertion `I != F.DeclRemap.end() && "Invalid index into decl index remap"' failed.`
- `llvm-project/clang/lib/Frontend/Rewrite/FrontendActions.cpp:231: virtual void clang::RewriteIncludesAction::RewriteImportsListener::visitModuleFile(StringRef, serialization::ModuleKind): Assertion `MF && "missing module file for loaded module?"' failed.`

Those two assertions happen after calls to `ModuleManager::lookupByFileName` and the bit added to `clang/lib/Frontend/Rewrite/FrontendActions.cpp` in this diff.

After I comment out these pieces, the toolchain compiles correctly, and the tests all pass.

In the non-asserts toolchain, the number of failing tests is around 25, but in that case the failures are segmentation faults (starting in `RewriteImportsListener::visitModuleFile`) and `fatal error: malformed or corrupted AST file: 'SourceLocation remap refers to unknown module, cannot find <path>`.

I am not saying I fully understand what the code is trying to do. I was just trying to understand if the `ext4` problem the comments talk about does or not apply to other filesystems. If I understand your explanations correctly, Darwin can opt-out and gets some benefit from not doing the double work (like absolute and relative paths being compared equal), while in other cases, the solution should just work (in Darwin it would work, but it is extra work). It seems, from my experience, that in some cases the extra work is not working in some systems.

**Comment by cachemeifyoucan** [comment]
@drodriguez Please consider this fix plus the previous commit that adds the workaround in `lookupModuleFile` as a complete entity. Please take all of them or take none of them. Any other configuration doesn't work.
